dev/src/collection-group.ts

```diff
@@ -81,48 +82,53 @@ export class CollectionGroup<
   async *getPartitions(
     desiredPartitionCount: number
   ): AsyncIterable<QueryPartition<AppModelType, DbModelType>> {
-    validateInteger('desiredPartitionCount', desiredPartitionCount, {
-      minValue: 1,
-    });
-
-    const tag = requestTag();
-    await this.firestore.initializeIfNeeded(tag);
-
     const partitions: Array<api.IValue>[] = [];
 
-    if (desiredPartitionCount > 1) {
-      // Partition queries require explicit ordering by __name__.
-      const queryWithDefaultOrder = this.orderBy(FieldPath.documentId());
-      const request: api.IPartitionQueryRequest =
-        queryWithDefaultOrder.toProto();
-
-      // Since we are always returning an extra partition (with an empty endBefore
-      // cursor), we reduce the desired partition count by one.
-      request.partitionCount = desiredPartitionCount - 1;
-
-      const stream = await this.firestore.requestStream(
-        'partitionQueryStream',
-        /* bidirectional= */ false,
-        request,
-        tag
-      );
-      stream.resume();
-
-      for await (const currentCursor of stream) {
-        partitions.push(currentCursor.values ?? []);
+    await this._firestore._traceUtil.startActiveSpan(
+      SPAN_NAME_PARTITION_QUERY,
+      async () => {
+        validateInteger('desiredPartitionCount', desiredPartitionCount, {
+          minValue: 1,
+        });
+
+        const tag = requestTag();
+        await this.firestore.initializeIfNeeded(tag);
+
+        if (desiredPartitionCount > 1) {
+          // Partition queries require explicit ordering by __name__.
+          const queryWithDefaultOrder = this.orderBy(FieldPath.documentId());
+          const request: api.IPartitionQueryRequest =
+            queryWithDefaultOrder.toProto();
+
+          // Since we are always returning an extra partition (with an empty endBefore
+          // cursor), we reduce the desired partition count by one.
+          request.partitionCount = desiredPartitionCount - 1;
+
+          const stream = await this.firestore.requestStream(
+            'partitionQueryStream',
+            /* bidirectional= */ false,
+            request,
+            tag
+          );
+          stream.resume();
+
+          for await (const currentCursor of stream) {
+            partitions.push(currentCursor.values ?? []);
+          }
+        }
+
+        logger(
+          'Firestore.getPartitions',
+          tag,
+          'Received %d partitions',
+          partitions.length
+        );
+
+        // Sort the partitions as they may not be ordered if responses are paged.
+        partitions.sort((l, r) => compareArrays(l, r));
       }
-    }
-
-    logger(
-      'Firestore.getPartitions',
-      tag,
-      'Received %d partitions',
-      partitions.length
     );
 
-    // Sort the partitions as they may not be ordered if responses are paged.
-    partitions.sort((l, r) => compareArrays(l, r));
-
     for (let i = 0; i < partitions.length; ++i) {
```

Comment:
why the code from 132-149 is not wrapped into the `startActiveSpan` cb function?

Reply:
this function (`getPartitions`) is an `async*` which generates multiple results asynchronously (`yield`s results) (line 133 and 143). This is not supported in `startActiveSpan` right now, but it's not a concern given that this function is simply yielding a new object in each loop iteration without doing any expensive work (for example, no RPCs or i/o or anything). All of the real work (`initializeIfNeeded`, `requestStream`, etc) is already done above inside the span. So the current span captures the time near-perfectly.

In fact I think this function could have just returned a `Promise<QueryPartition<AppModelType, DbModelType>>` rather than `AsyncIterable<QueryPartition<AppModelType, DbModelType>>`